File: Frontend/PointCloudProcessing/open3d/downsampling.py
import open3d as o3d
import numpy as np

def load_kitti_bin_file(bin_file):
    """
    Load point cloud from a KITTI .bin file.

    Args:
        bin_file (str): Path to the .bin file.

    Returns:
        o3d.geometry.PointCloud: Open3D PointCloud object.
    """
    # Each point in the .bin file is structured as [x, y, z, reflectance]
    point_cloud = np.fromfile(bin_file, dtype=np.float32).reshape(-1, 4)
    points = point_cloud[:, :3]  # Extract x, y, z coordinates
    # Create Open3D PointCloud
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points)
    return pcd

def downsample_point_cloud(pcd, voxel_size):
    """
    Downsample the point cloud using voxel downsampling.

    Args:
        pcd (o3d.geometry.PointCloud): Original point cloud.
        voxel_size (float): Voxel size for downsampling.

    Returns:
        o3d.geometry.PointCloud: Downsampled point cloud.
    """
    downsampled_pcd = pcd.voxel_down_sample(voxel_size=voxel_size)
    return downsampled_pcd

def visualize_point_cloud(pcd, color, point_size=1.0, window_name='Point Cloud'):
    """
    Visualize a single point cloud.

    Args:
        pcd (o3d.geometry.PointCloud): Point cloud to visualize.
        color (list): RGB color for the point cloud.
        point_size (float): Size of the points in the visualization.
        window_name (str): Name of the visualization window.
    """
    # Assign color to point cloud
    pcd.paint_uniform_color(color)

    # Create a Visualizer object
    vis = o3d.visualization.Visualizer()
    vis.create_window(window_name=window_name, width=800, height=600)

    # Add point cloud to the visualizer
    vis.add_geometry(pcd)

    # Get the render options and set the point size
    render_option = vis.get_render_option()
    render_option.point_size = point_size  # Adjust point size as needed

    # Run the visualizer
    vis.run()
    vis.destroy_window()

if __name__ == "__main__":
    # Path to the KITTI .bin file
    bin_file = '../../../../KITTI/lidar/bin/000004.bin'

    # Load the original point cloud
    original_pcd = load_kitti_bin_file(bin_file)

    # Downsample the point cloud
    voxel_size_1 = 0.2  # Adjust the voxel size as needed
    downsampled_pcd = downsample_point_cloud(original_pcd, voxel_size=voxel_size_1)

    # Double downsample the point cloud (downsample the downsampled point cloud)
    voxel_size_2 = 0.4  # Adjust the voxel size as needed
    double_downsampled_pcd = downsample_point_cloud(downsampled_pcd, voxel_size=voxel_size_2)

    # Optionally, print the number of points in each point cloud
    print(f"Original point cloud has {len(original_pcd.points)} points.")
    print(f"First downsampled point cloud has {len(downsampled_pcd.points)} points.")
    print(f"Second downsampled point cloud has {len(double_downsampled_pcd.points)} points.")

    # Visualize each point cloud separately
    # Original point cloud
    visualize_point_cloud(original_pcd, color=[1, 0, 0], point_size=3.0, window_name='Original Point Cloud')

    # First downsampled point cloud
    visualize_point_cloud(downsampled_pcd, color=[0, 1, 0], point_size=3.0, window_name='First Downsampled Point Cloud')

    # Second downsampled point cloud
    visualize_point_cloud(double_downsampled_pcd, color=[0, 0, 1], point_size=3.0, window_name='Second Downsampled Point Cloud')

File: Frontend/PointCloudProcessing/PCL/downsampling.cpp
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <pcl/visualization/pcl_visualizer.h>
#include <pcl/filters/voxel_grid.h>
#include <fstream>
#include <vector>

pcl::PointCloud<pcl::PointXYZ>::Ptr loadKittiBinFile(const std::string& binFile) {
    // Create point cloud object
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    
    // Read binary file
    std::ifstream file(binFile, std::ios::binary);
    if (!file.is_open()) {
        std::cerr << "Could not open file: " << binFile << std::endl;
        return cloud;
    }

    // Get file size
    file.seekg(0, std::ios::end);
    const size_t num_elements = file.tellg() / (4 * sizeof(float));
    file.seekg(0, std::ios::beg);

    // Read point cloud data
    std::vector<float> buffer(num_elements * 4);
    file.read(reinterpret_cast<char*>(buffer.data()), buffer.size() * sizeof(float));
    
    // Convert to PCL point cloud
    cloud->points.reserve(num_elements);
    for (size_t i = 0; i < num_elements; i++) {
        pcl::PointXYZ point;
        point.x = buffer[i * 4];
        point.y = buffer[i * 4 + 1];
        point.z = buffer[i * 4 + 2];
        // Note: we ignore reflectance (buffer[i * 4 + 3])
        cloud->points.push_back(point);
    }
    
    cloud->width = cloud->points.size();
    cloud->height = 1;
    cloud->is_dense = true;
    
    return cloud;
}

pcl::PointCloud<pcl::PointXYZ>::Ptr downsamplePointCloud(
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud,
    float voxelSize) {
    
    pcl::PointCloud<pcl::PointXYZ>::Ptr downsampled_cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::VoxelGrid<pcl::PointXYZ> voxel_filter;
    voxel_filter.setInputCloud(cloud);
    voxel_filter.setLeafSize(voxelSize, voxelSize, voxelSize);
    voxel_filter.filter(*downsampled_cloud);
    
    return downsampled_cloud;
}

void visualizePointCloud(
    const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud,
    const std::vector<double>& color,
    double pointSize,
    const std::string& windowName) {
    
    // Check if cloud is valid and not empty
    if (!cloud || cloud->empty()) {
        std::cerr << "Warning: Empty or invalid point cloud for " << windowName << ". Skipping visualization." << std::endl;
        return;
    }
    
    // Create visualizer
    pcl::visualization::PCLVisualizer::Ptr viewer(new pcl::visualization::PCLVisualizer(windowName));
    viewer->setBackgroundColor(0, 0, 0);
    
    // Add point cloud to viewer
    pcl::visualization::PointCloudColorHandlerCustom<pcl::PointXYZ> single_color(
        cloud, color[0] * 255, color[1] * 255, color[2] * 255);
    viewer->addPointCloud<pcl::PointXYZ>(cloud, single_color, "cloud");
    viewer->setPointCloudRenderingProperties(
        pcl::visualization::PCL_VISUALIZER_POINT_SIZE, pointSize, "cloud");
    
    // Wait until viewer is closed
    while (!viewer->wasStopped()) {
        viewer->spinOnce(100);
    }
}

int main(int argc, char** argv) {
    // Path to the KITTI .bin file
    std::string binFile = "../../../../../KITTI/lidar/bin/000004.bin";
    
    // Load the original point cloud
    auto originalCloud = loadKittiBinFile(binFile);
    
    // Check if cloud loading was successful
    if (originalCloud->empty()) {
        std::cerr << "Failed to load point cloud. Exiting." << std::endl;
        return -1;
    }
    
    // Downsample the point cloud
    float voxelSize1 = 0.2f;
    auto downsampledCloud = downsamplePointCloud(originalCloud, voxelSize1);
    
    // Double downsample the point cloud
    float voxelSize2 = 0.4f;
    auto doubleDownsampledCloud = downsamplePointCloud(downsampledCloud, voxelSize2);
    
    // Print point cloud sizes
    std::cout << "Original point cloud has " << originalCloud->size() << " points." << std::endl;
    std::cout << "First downsampled point cloud has " << downsampledCloud->size() << " points." << std::endl;
    std::cout << "Second downsampled point cloud has " << doubleDownsampledCloud->size() << " points." << std::endl;
    
    // Visualize point clouds
    // Original point cloud (red)
    visualizePointCloud(originalCloud, {1.0, 0.0, 0.0}, 3.0, "Original Point Cloud");
    
    // First downsampled point cloud (green)
    visualizePointCloud(downsampledCloud, {0.0, 1.0, 0.0}, 3.0, "First Downsampled Point Cloud");
    
    // Second downsampled point cloud (blue)
    visualizePointCloud(doubleDownsampledCloud, {0.0, 0.0, 1.0}, 3.0, "Second Downsampled Point Cloud");
    
    return 0;
}